src/coreclr/jit/gentree.cpp

```diff
@@ -11702,6 +11702,66 @@ GenTree* Compiler::gtFoldExprCall(GenTreeCall* call)
     return call;
 }
 
+#ifdef FEATURE_HW_INTRINSICS
+
+//------------------------------------------------------------------------
+// gtFoldHWIntrinsic: see if a HWIntrinsic is foldable
+//
+// Arguments:
+//    node - HWIntrinsic node to examine
+//
+// Returns:
+//    The original node if no folding happened.
+//    An alternative tree if folding happens.
+//
+// Notes:
+//    Checks for HWIntrinsic nodes to Vector64.Create/Vector128.Create/Vector256.Create,
+//    and if the call is to one of these, attempt to optimize.
+
+GenTree* Compiler::gtFoldHWIntrinsic(GenTreeHWIntrinsic* node)
+{
+    // Defer folding if not optimizing.
+    if (opts.OptimizationDisabled())
+    {
+        return node;
+    }
+
+    switch (node->GetHWIntrinsicId())
+    {
+        case NI_Vector128_Create:
+#if defined(TARGET_XARCH)
+        case NI_Vector256_Create:
+#elif defined(TARGET_ARM64)
+        case NI_Vector64_Create:
+#endif
+        {
+            bool hwAllArgsAreConstZero = true;
+            for (GenTree* arg : node->Operands())
+            {
+                if (!arg->IsIntegralConst(0) && !arg->IsFloatPositiveZero())
+                {
+                    hwAllArgsAreConstZero = false;
+                    break;
+                }
+            }
+
+            if (hwAllArgsAreConstZero)
+            {
+                return gtNewSimdZeroNode(node->gtType, node->GetSimdBaseJitType(), node->GetSimdSize(),
+                                         node->IsSimdAsHWIntrinsic());
+            }
+            break;
```

Comment:
This transform deletes nodes, thus needs to be done under `!optValnumCSE_phase` guard (check it at the call site and `assert(!optValnumCSE_phase)` in the method - we're expecting to stop running most "expression optimizers" under CSE in the future).

This method, presently, loses a VN on tree. It can (should) preserve the VN instead:
```c
var zero = gtNew{...};
zero->SetVNsFromNode(node);

return zero;
```
If it will not, it needs to be under `fgGlobalMorph`.

Reply:
I'm not familiar with value numbering; I will try to look for examples where this happens.

Reply:
Note that what I said above is a strict correctness requirement on morphing functions: they must not delete nodes under CSE (technically they must not delete active CSE candidates, but for simplicity and robustness it is preferred to simply not run folding under CSE), they must preserve VNs, not doing so can (and has) lead to hard to diagnose silent bad codegen bugs and/or crashes. Note that I think it will be hard to reproduce something going wrong here (perhaps with aggressive CSE for constants...), I would not recommend spending time on finding a reproduction.

There is a possible exception to the "must preserve VNs" rule in that it should be "fine" to create new nodes with `NoVN`, but we shy away from doing so for simplicity purposes (it is easier to reason about - "Does morph preserve VNs?" -- "Yes, always", instead of "Well, sort of kind of").

Reply:
Makes sense to me, thank you for the explanation. I'm all for the lens of a strict correctness perspective, so this helps a lot. I need to spend more time understanding how this mechanism works when the tree is changed.

Reply:
There is also this call I found:
```cpp
SetOper(oper, GenTree::PRESERVE_VN);
```

Reply:
I think we don't need to preserve anything since we do not do the optimization in the CSE phase.

Reply:
It is the case that morph should always preserve VNs as it can be called from any optimization phase. The most obvious example where that can be detrimental is hoisting, where we morph the hoisted tree, and if it loses its VN, CSE won't be able to actually remove the tree inside the loop.

I see the optimization uses `ResetHWIntrinsicId` now, which preserves the VN, so it is of no concern anymore.

Reply:
Yup, `ResetHWIntrinsicId` fit naturally in this case. Thank you for the help!